src/pages/RealtimeMapPage.tsx

```diff
@@ -60,66 +65,87 @@ function numberToColorHsl(i: number, max: number) {
-        formatTime(to) || '2023-05-01T12:01:00+02:00'
-      }&order_by=id%20asc&limit=${limit}`
-      const response = await fetch(url)
-      const data = await response.json()
-      const points: Point[] = data.map((point: { lat: number; lon: number; velocity: number }) => ({
-        loc: [point.lat, point.lon],
-        color: point.velocity,
-        point,
-      }))
-      setPositions(points)
-      setLoaded(points.length)
-    })()
-  }, [limit])
+  const locations = useVehicleLocations({
+    from: formatTime(from),
+    to: formatTime(to),
+  })
+
+  const loaded = locations.length
+
+  const positions = useMemo(() => {
+    let pos = locations.map<Point>((location) => ({
+      loc: [location.lat, location.lon],
+      color: location.velocity,
+      operator: location.siri_route__operator_ref,
+      bearing: location.bearing,
+      recorded_at_time: new Date(location.recorded_at_time).getTime(),
+      point: location,
+    }))
+    // keep only the latest point for each vehicle
+    pos = pos.filter((p) =>
+      pos.every(
+        (p2) =>
+          p2.point!.siri_ride__vehicle_ref !== p.point!.siri_ride__vehicle_ref ||
+          p2.recorded_at_time! <= p.recorded_at_time!,
+      ),
+    )
+    return pos
+  }, [locations])
 
   return (
     <div className="map-container">
       <div className="map-header">
         <h1>Realtime Map</h1>
-        <div className="map-header-buttons">
-          <button onClick={() => setLimit(500)}>500</button>
-          <button onClick={() => setLimit(1000)}>1000</button>
-          <button onClick={() => setLimit(5000)}>5000</button>
-          <button onClick={() => setLimit(10000)}>10000</button>
-          <button onClick={() => setLimit(10000)}>30000</button>
-        </div>
         <div className="map-header-buttons">
           <label>
             מתאריך
-            <input type="datetime-local" value={from} onChange={(e) => setFrom(e.target.value)} />
+            <input
+              type="datetime-local"
+              value={from.slice(0, 16)}
+              onChange={(e) => {
+                setFrom(e.target.value)
+                setTo(formatTime(+new Date(e.target.value) + (+new Date(to) - +new Date(from))))
+              }}
+            />
           </label>{' '}
           {` `}
           <label>
-            עד תאריך
-            <input type="datetime-local" value={to} onChange={(e) => setTo(e.target.value)} />
+            צפה במיקומי אוטובוסים בטווח של
+            <input
+              type="number"
+              value={(+new Date(to) - +new Date(from)) / 1000 / 60}
+              onChange={(e) => setTo(formatTime(+new Date(from) + +e.target.value * 1000 * 60))}
+            />
+            דקות
           </label>
         </div>
+        <div className="map-header-buttons">
+          <button
+            onClick={() => {
+              setFrom(formatTime(+new Date() - 5 * 1000 * 60))
+              setTo(formatTime(+new Date() - 4 * 1000 * 60))
```

Comment:
would be really nice to use some helper methods or better yet a library for these converts, they're not readable

Reply:
I don't feel it's not readable - I'll add a comment anyway